implementation of new orographic drag parameterization scheme to alleviate wind bias in E3SM

jinboxie opened the pull request:
Add a new suite of orographic drag parameterization schemes into EAM. 
It includes 4 components all combined in one module (i.e. subroutine gwdo2d). 
The schemes include orographic gravity wave drag (oGWD, Xie et al.,2020), 
flow-blocking drag (FBD, Xie et al.,2020), small-scale GWD (sGWD, Tsiringakis et al.,2017), 
and turbulent-scale orographic form drag (TOFD, Beljaars et al., 2004).

The scheme is distributed by the interface named gwdo_gsd. The oGWD, FBD, and 
sGWD are implemented in gw_drag.F90, while TOFD is implemented in clubb_intr.F90 
to join the vertical diffusion process. Another change is in the startup_initialconds.F90 
and gw_drag.F90 where the new topographic parameters (OA, OC, OL) are input into the model.

Although it is possible to turn each new scheme on/off individually, the only validated 
config is to have them all on (and the original scheme off).

Technical  guide for this scheme is available at  
https://docs.e3sm.org/E3SM/EAM/tech-guide/orodrag/

[BFB]

-------

(formation of the additional topographic parameters are documented [here](https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/4625563753/Generation+of+the+additional+topographic+parameters+for+the+new+orographic+drag+scheme+considering+orographic+anisotropy+for+all+flow+directions)).

The results tested in E3SMv2/v3 documented [here](https://acme-climate.atlassian.net/wiki/spaces/ATMOS/pages/4710662196/Testing+simulation+for+new+suite+of+orographic+drag+schemes+in+E3SM). The overall 5-yr F2010 results between CTL run (default E3SMv3) and 3GWD-3FBD-sGWD-TOFD run (experimental run using new schemes) show that the new schemes help to alleviate the 10-m wind speed bias compared to ERA5 in E3SMv3.

The coupled simulation has run past over 10 years and the results are documented [here](https://acme-climate.atlassian.net/wiki/spaces/ATMOS/pages/4731174922/Coupled+simulation+for+testing+of+the+new+orographic+drag+schemes). 

-------

The namelist control is added for the 4 schemes. The 4 schemes are separately control by namelist variables 
use_od_ls (oGWD)
use_od_bl (FBD)
use_od_ss (sGWD)
use_od_fd (TOFD)
tunable parameters in the schemes are also controllable via namelist, e.g.
od_ls_ncleff (oGWD)
od_bl_ncd (FBD)
od_ss_sncleff (sGWD).

-------

The code has passed all the tests in the "e3sm_atm_developer" & "e3sm_atm_integration" suites with new scheme closed and topo file modified.
The code has also passed all the test in the "e3sm_atm_orodrag" where the v3 default GWD is turned off (use_gw_oro=.false.) and the new schemes turned on 
(use_od_ls=.true.
 use_od_bl=.true.
 use_od_ss=.true.
 use_od_fd=.true.).
The e3sm_atm_orodrag is a suite of new tests for development of the new schemes in e3sm.

-------

<img width="656" alt="image" src="https://github.com/user-attachments/assets/90b63dba-2547-4ee8-b7cd-70864d5e3f98">


-------


 ![image](https://github.com/user-attachments/assets/3ff5b39f-3ba3-4533-b231-25483a4ff3ff)
Figure 1 5-yr F2010 simulation for E3SMv3 (0001-0005) vs ERA5 climatology (1979-2019) for 10-m wind speed (a-b) and 2-m temperature (c-d).

-------
Further task lists before merging

- [x] Do a 10-yr long coupled run.
- [x] Add warnings for selections of drag schemes in the model.
- [x] Separate the topo data generation workflow from the original topo tools in E3SM
- [x] Pass the stealth test for orographic drag schemes.
- [x] Test if the new topo data may impact the results compared with default E3SM baseline.

oksanaguba commented:
A few questions, please -- are these drags energy conserving? I assume at least the one in clubb is not (from reading the code)? Should we run some checks to confirm this?

Separately, if these are topo drag parameterizations, why does this PR introduce changes in the topo tool? In other words, how/why are these parameterizations affecting topo generation?

Were these parameterizations tuned to show smaller bias? I assume only one was used to report smaller biases as above, what about the other 3? (unless they all run together?)

If i knew that this code base (physpkg.F90, etc) is to be used for a while, i would argue for you to remove the code from clubb_intr and make is a separate module/routine. You can then use that new code to add another forcing term to be "absorbed/diffused" by clubb. This is to separate clubb interface code from clearly a different parameterization.

thanks!

jinboxie commented:
> A few questions, please -- are these drags energy conserving? I assume at least the one in clubb is not (from reading the code)? Should we run some checks to confirm this?
> 
> Separately, if these are topo drag parameterizations, why does this PR introduce changes in the topo tool? In other words, how/why are these parameterizations affecting topo generation?
> 
> Were these parameterizations tuned to show smaller bias? I assume only one was used to report smaller biases as above, what about the other 3? (unless they all run together?) UPDATE -- it seems they ran together.
> 
> If i knew that this code base (physpkg.F90, etc) is to be used for a while, i would suggest for you to remove the code from clubb_intr and make is a separate module/routine. You can then use that new code to add another forcing term to cam_in structure to be "absorbed/diffused" by clubb. This is to separate clubb interface code from clearly a different parameterization.
> 
> thanks!

Hi Oksana,
Thank you for the questions! 
In terms of the energy conserving, I'm not sure I understand this correctly. Did you mean the fixer that gw energy fixer that was implemented in gw_drag.F90? I'm not familiar with this part, could you maybe clarify on how to test it?

The introduction of changes in topo tool does not affect the current topo generation. It is used to generate additional topographic parameters (such as orographic asymmetry, effective orographic length) that are needed for the new parameterization. The process is documented [here](https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/4625563753/Generation+of+the+additional+topographic+parameters+for+the+new+orographic+drag+scheme+considering+orographic+anisotropy+for+all+flow+directions).

For implementation of turbulent orographic form drag (TOFD, the TMS equivalent) into the CLUBB intr. I was following the previous implementation of TMS. What we input into the clubb_intr is an interface that utilizes the module of new schemes (4 schemes modulize together, and easily distributed through open and closing flags). 

Unlike TMS, the TOFD can generate a profile rather than a single-level surface stress. We had to find a way to input this forcing profile of TOFD into the clubb to join the vertical diffusion (or maybe what you meant by "absorbed" by clubb?). 
So we are currently following TMS's positioning in clubb_intr, the minor difference is that TMS is inputting surface wind forcing (upwp_sfc, vpwp_sfc) while we are inputting wind profile forcing terms (um_forcing, vm_forcing) into clubb at the clubb_intr.

So in this code, I was inputting an interface for the calculation of the TOFD into clubb_intr while the module is somewhere else in physics folder. I was wondering if you'd meant to put this interface of the new scheme out in physpkg.F90 instead?

jinboxie commented:
@mt5555 
Hi Mark,
I have a question in terms of the topography file. The current implemented scheme involves use of additional topographic parameters that are added in the topo file, (generation method documented [here](https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/4625563753/Generation+of+the+additional+topographic+parameters+for+the+new+orographic+drag+scheme+considering+orographic+anisotropy+for+all+flow+directions)). Take ne30pg2 as an example (USGS-gtopo30_ne30np4pg2_x6t-SGH_forOroDrag.c20241001.nc in replacement of USGS-gtopo30_ne30np4pg2_x6t-SGH.c20210614.nc), the original variables such as PHIS, SGH, SGH30 are the same, except for addition of parameters such as orographic asymmetry (OA), orographic convexity (OC), etc. These are needed when new scheme is open while not needed when the default GWD is used.
I was wondering if I can update the file in the namelist like this (the topo data is on chrysalis input directory and shared over other servers), or would you suggest alternative options?
Jinbo

rljacob commented:
Someone needs to approve this.

whannah1 commented on components/eam/bld/namelist_files/namelist_defaults_eam.xml:
```diff
@@ -126,14 +126,13 @@
 <bnd_topo hgrid="ne16np4"  npg="2">atm/cam/topo/USGS-gtopo30_ne16np4pg2_16xdel2_20200527.nc</bnd_topo>
 <bnd_topo hgrid="ne30np4"  npg="0">atm/cam/topo/USGS-gtopo30_ne30np4_16xdel2-PFC-consistentSGH.nc</bnd_topo>
-<bnd_topo hgrid="ne30np4"  npg="2">atm/cam/topo/USGS-gtopo30_ne30np4pg2_x6t-SGH.c20210614.nc</bnd_topo>
+<bnd_topo hgrid="ne30np4"  npg="2">atm/cam/topo/USGS-gtopo30_ne30np4pg2_x6t-SGH_forOroDrag.c20241001.nc</bnd_topo>
```
This seems to be turning the scheme on by default, which makes it non-BFB right? I didn't a BFB tag on the PR.

whannah1 commented on components/eam/src/physics/cam/clubb_intr.F90:
```diff
@@ -1155,7 +1163,10 @@ subroutine clubb_tend_cam( &
+   use gw_common,          only: grid_size,gw_oro_interface
+   use hycoef,             only: etamid
+   use physconst,          only: rh2o,pi,rearth,r_universal
+   !!get the znu,znw,p_top set to 0
```
What is this comment referring to?

whannah1 commented on components/eam/src/physics/cam/clubb_intr.F90:
```diff
@@ -3136,13 +3207,17 @@ subroutine clubb_surface (state, cam_in, ustar, obklen)
     real(r8) :: kinheat                                         ! kinematic surface heat flux
     real(r8) :: kinwat                                          ! kinematic surface vapor flux
     real(r8) :: kbfs                                            ! kinematic surface buoyancy flux
+    real(r8) :: kbfs_pcol(pcols)
```
This variable needs a descriptive comment

whannah1 commented on components/eam/src/physics/cam/clubb_intr.F90:
```diff
@@ -3173,14 +3248,44 @@ subroutine clubb_surface (state, cam_in, ustar, obklen)
        end if
     enddo
-
+    !
```
Why are you adding these pointless comment lines?

whannah1 commented on components/eam/src/physics/cam/clubb_intr.F90:
```diff
@@ -3173,14 +3248,44 @@ subroutine clubb_surface (state, cam_in, ustar, obklen)
+         th_lv(i,k) = state%t(i,k)*state%exner(i,k)
+         if (use_sgv) then
+           thv_lv(i,k) = th_lv(i,k)*(1.0_r8+zvir*state%q(i,k,ixq) &
+                    - state%q(i,k,ixcldliq))  !PMA corrects thv formula
```
is "PMA" for someone's initials? We don't want tags like this in the code.

whannah1 replied:
Also - a better description about what is being fixed would be better since it's very unclear what the problem was.

whannah1 commented on components/eam/src/physics/cam/comsrf.F90:
```diff
@@ -134,4 +140,32 @@ subroutine initialize_comsrf
+        allocate (oc(pcols,begchunk:endchunk))
+        allocate (oadir(pcols,nvar_dirOA,begchunk:endchunk))
+        allocate (ol(pcols,nvar_dirOL,begchunk:endchunk))
+        var(:,:)=nan
+        var30(:,:)=nan
+        oc    (:,:) = nan
+        oadir (:,:,:) = nan
+        ol  (:,:,:) = nan
```
Clean up the vertical alignment here

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -741,5 +745,1387 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+  real(r8) :: dusfc_ss_local(pcols)
+  real(r8) :: dvsfc_ss_local(pcols)
+  real(r8) :: dusfc_fd_local(pcols)
+  real(r8) :: dvsfc_fd_local(pcols)
+
+                !
+                ncol=state%ncol
+                !convert heights above surface to heights above sea level
+                !obtain z,dz,dx,dy
+                !ztop and zbot are already reversed, start from bottom to top
+                kpbl2d_in=0_r8
```
Can you make the vertical alignment consistent? Breaks in the alignment like this are one of my all-time biggest pet peeves, and have led me personally to incorrectly interpret code, so I can't stand sloppy code like this.

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -741,5 +745,1387 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+pblh_get_level_idx = -1
+found=.False.
+
+do i = 1, pver
+        if((pblheight >= height_array(i+1).and.pblheight <height_array(i)))then
+                pblh_get_level_idx =  pver+1-i
```
Here's another example of weird vertical alignment - why are you not indenting the all lines of the routine?

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -741,5 +745,1387 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+  real(r8),     dimension( ims:ime, kms:kme)   ::                                    &
+                                  dtaux2d_ls,dtauy2d_ls,dtaux2d_bl,dtauy2d_bl, &
+                                  dtaux2d_ss,dtauy2d_ss,dtaux2d_fd,dtauy2d_fd
+  real(r8),      dimension( ims:ime, kms:kme )                          , &
+
+             intent(in   )   ::                        u3d, &
+                                                                v3d
+!
+  integer,   dimension( ims:ime )                                   , &
+             intent(in  )   ::                             kpbl2d
+
+  real(r8),   dimension( ims:ime )                                      , &
+        intent(in  )   ::                                   pblh, &
+                                                                 br, &
```
Again, why is the vertical alignment so sloppy?

whannah1 commented on components/eam/src/physics/cam/gw_drag.F90:
```diff
@@ -24,7 +24,8 @@ module gw_drag
   use shr_kind_mod,  only: r8 => shr_kind_r8
-  use ppgrid,        only: pcols, pver
+  use ppgrid,        only: pcols,pver,pverp,nvar_dirOA,nvar_dirOL,begchunk,endchunk
+  use hycoef,             only: hyai, hybi, hyam, hybm, etamid !get the znu,znw,p_top set to 0
```
I'm confused how this comment relates to the line it's currently on?

whannah1 commented on components/eam/src/physics/cam/gw_drag.F90:
```diff
@@ -893,11 +990,44 @@ subroutine gw_tend(state, sgh, pbuf, dt, ptend, cam_in)
-     ! Compute the temperature tendency from energy conservation
-     ! (includes spectrum).
-
+  endif
+        ! Add the orographic tendencies to the spectrum tendencies
+        ! Compute the temperature tendency from energy conservation
+        ! (includes spectrum).
+        ! both old and new gwd scheme will add the tendency to circulation
+  if (use_gw_oro.or.    &
```
Why do these comments have extra indentation?

whannah1 commented on components/eam/src/physics/cam/physpkg.F90:
```diff
@@ -1321,15 +1321,15 @@ subroutine phys_run2(phys_state, ztodt, phys_tend, pbuf2d,  cam_out, &
 
     use cam_diagnostics,only: diag_deallocate, diag_surf
-    use comsrf,         only: trefmxav, trefmnav, sgh, sgh30, fsds 
+    use comsrf,         only: trefmxav, trefmnav, sgh, sgh30, fsds, var, var30,oc,oadir,ol
```
Why the inconsistent spacing!? All new code should match the code around it!

whannah1 commented on components/eam/src/physics/cam/physpkg.F90:
```diff
@@ -1432,7 +1432,15 @@ subroutine phys_run2(phys_state, ztodt, phys_tend, pbuf2d,  cam_out, &
-
+       ! for tranport of ogwd related parameters
+       if (use_od_ls.or.use_od_bl) then
+       phys_state(c)%var(:)=var(:,c)                   
```
indent

whannah1 commented on components/eam/tools/topo_tool/orographic_drag_toolkit/cube_to_target.F90:
```diff
@@ -0,0 +1,2550 @@
+!
```
Why are we adding a second version of cube_to_target? 
This seems like a recipe for diverging versions where one is working and the other is broken, right?
I feel like we should only ever have a single version.

jinboxie replied:
This is a suite of toolkit for generation of additional topographic parameters for the new scheme (documented [here](https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/4625563753/Generation+of+the+additional+topographic+parameters+for+the+new+orographic+drag+scheme+considering+orographic+anisotropy+for+all+flow+directions)).
The derivation code is developed before also based on cube_to_target. We'd want to make an independent workflow so as to be separate from the workflow now maintained by Mark's @mt5555.
As documented in the confluence ([here](https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/4625563753/Generation+of+the+additional+topographic+parameters+for+the+new+orographic+drag+scheme+considering+orographic+anisotropy+for+all+flow+directions)), we'd generate the additional topographic parameters and append them to the topo data that was generated by the default topo workflow in E3SM.

whannah1 replied:
you should name the tool something else then - having two tools with the same name is definitely going to lead to confusion.

But also - if these quantities need to be appending to a topo file, why not enhance the existing cube_to_target tool to produce the quantities and not make more steps for people to work through when they create new grids? This seems like it's going to make everything harder to maintain rather than working within the existing workflows.

jinboxie commented:
> Someone needs to approve this.

Hi Rob @rljacob , the current code is testing underway, and should be finished soon. We are waiting for a 10-yrs coupled testing run before merging. Thanks!

rljacob commented:
@jinboxie is this PR BFB ?

jinboxie commented:
Hi Walter @whannah1 ,
Thank you for your comments on the code. I have modified accordingly, specifically on the formatting such as the vertical alignment you mentioned. Please see if you have further comments on the latest commit. Thanks!
Best,
Jinbo

jinboxie commented:
HI Mark @mt5555 ,
It is tested that the results using the new topo data generated is BFB to the default E3SM simulations. I've also conducted over 10-yr coupled simulations and it runs fine. Please see if you'd have further comments.
Jinbo

whannah1 commented on components/eam/bld/namelist_files/namelist_definition.xml:
```diff
@@ -1078,6 +1078,48 @@ Whether or not to enable GWD brute-force energy fix.
+       group="phys_ctl_nl" valid_values="" >
+Whether or not to enable nonlinear orographic gravity wave drag (oGWD).
+Default: set by build-namelist.
+</entry>
+
+<entry id="use_od_bl" type="logical" category="gw_drag"
+       group="phys_ctl_nl" valid_values="" >
+Whether or not to enable flow-blocking drag (FBD).
+Default: set by build-namelist.
+</entry>
+
+<entry id="use_od_ss" type="logical" category="gw_drag"
+       group="phys_ctl_nl" valid_values="" >
+Whether or not to enable small-scale orographic GWD drag (sGWD).
+Default: set by build-namelist.
+</entry>
+
+<entry id="use_od_fd" type="logical" category="pbl"
+       group="phys_ctl_nl" valid_values="" >
+Whether or not to enable turbulent orographic form drag (TOFD).
+Default: set by build-namelist.
+</entry>
+
+<entry id="ncleff_ls" type="real" category="gw_drag"
+       group="oro_drag_nl" valid_values="" >
+Tuning parameter of orographic GWD (oGWD). See <varname>use_od_ls</varname>.
+Default: set by build-namelist.
+</entry>
+
+<entry id="ncd_bl" type="real" category="gw_drag"
+       group="oro_drag_nl" valid_values="" >
+Tuning parameter of flow-blocking drag (FBD). See <varname>use_od_bl</varname>.
+Default: set by build-namelist.
+</entry>
+
+<entry id="sncleff_ss" type="real" category="gw_drag"
+       group="oro_drag_nl" valid_values="" >
+Tuning parameter of small-scale GWD (sGWD). See <varname>use_od_ss</varname>.
+Default: set by build-namelist.
+</entry>
```
Sorry to be annoying - but these variable names are very unclear. There's no chance a user would be able to guess what they are used for if they were scanning the namelist. Could you rename these to be something more descriptive? Maybe they could all have the same prefix like "gw_oro_" to make it clear that they go together?

whannah1 replied:
I guess since the new schemes replace the old orographic GW scheme you probably want a prefix that differentiates the two, like "gw_oro_new_"?

jinboxie replied:
we now use od (orographic drag) for all the schemes. use_od_? are the series for schemes, and od_ls_cleff are for tunable parameters in od.

whannah1 commented on components/eam/src/control/startup_initialconds.F90:
```diff
@@ -5,16 +5,28 @@ module startup_initialconds
 save
 
 public :: initial_conds ! Read in initial conditions (dycore dependent)
+!added for orographic drag
```
Should this comment indicate that it's the "new" oro drag scheme for clarity?

whannah1 commented on components/eam/src/physics/cam/clubb_intr.F90:
```diff
@@ -927,7 +928,27 @@ subroutine clubb_ini_cam(pbuf2d, dp1_in)
-
+    !
+    if (use_od_fd) then
+    !!added for TOFD output
```
what's "TOFD"?

whannah1 replied:
I see it's defined below - I would spell it out here too

whannah1 commented on components/eam/src/physics/cam/clubb_intr.F90:
```diff
@@ -1946,7 +1991,36 @@ subroutine clubb_tend_cam( &
+        gwd_bl=.false.
+        gwd_ss=.false.
+        gwd_fd=use_od_fd
+        dummy_nm=0.0_r8
+        dummy_utgw=0.0_r8
+        dummy_vtgw=0.0_r8
+        dummy_ttgw=0.0_r8
```
What's the point of these dummy variables? Are they always going to be zero?

jinboxie replied:
Yes. Because we use interface of od_drags, to keep the inputs in same format, we use these dummy variables.

whannah1 commented on components/eam/src/physics/cam/clubb_intr.F90:
```diff
@@ -3181,6 +3268,41 @@ subroutine clubb_surface (state, cam_in, ustar, obklen)
                         kinheat, kinwat, kbfs, obklen(i) )
     enddo
 
+    if (use_od_ss) then
+    !add calculation of bulk richardson number here
+    !
+    !compute the whole level th and thv for diagnose of bulk richardson number
+    thv_lv=0.0_r8
+    th_lv=0.0_r8
+
+    !use the same virtual potential temperature formula as above (thv) except for all vertical levels
+    !used for bulk richardson number below in pblintd_ri
+    do i=1,ncol
+      do k=1,pver
+         th_lv(i,k) = state%t(i,k)*state%exner(i,k)
+         if (use_sgv) then
+           thv_lv(i,k) = th_lv(i,k)*(1.0_r8+zvir*state%q(i,k,ixq) &
+                    - state%q(i,k,ixcldliq))  
+         else
+           thv_lv(i,k) = th_lv(i,k)*(1.0_r8+zvir*state%q(i,k,ixq))
+         end if
+      enddo
+    enddo
+
+    !recalculate the kbfs stored in kbfs_pcol for bulk richardson number in pblintd_ri
+    kbfs_pcol=0.0_r8
+    do i=1,ncol
+        call calc_ustar( state%t(i,pver), state%pmid(i,pver), cam_in%wsx(i), cam_in%wsy(i), rrho, ustar(i) )
+        call calc_obklen( th(i), thv(i), cam_in%cflx(i,1), cam_in%shf(i), rrho, ustar(i), &
+                        kinheat, kinwat, kbfs, obklen(i) )
+        kbfs_pcol(i)=kbfs
+    enddo
+
+    !calculate the bulk richardson number
+    call pblintd_ri(ncol, gravit, thv_lv, state%zm, state%u, state%v, &
+                ustar, obklen, kbfs_pcol, state%ribulk)
+    endif
```
Here's another place with inconsistent indentation - I know we don't have an "official" policy on code indentation, but I think it's common sense that the code within an if block should be indented

whannah1 commented on components/eam/src/physics/cam/gw_drag.F90:
```diff
@@ -287,7 +296,41 @@ subroutine gw_init()
+  !
+  grid_id = cam_grid_id('physgrid')
+  !
+  if (use_od_ls.or.use_od_bl) then
+                if (.not. cam_grid_check(grid_id)) then
+                call endrun(trim(subname)//': Internal error, no "physgrid" grid')
+                end if
+                call cam_grid_get_dim_names(grid_id, dim1name, dim2name)
+                !!
+                call initialize_comsrf2()
+                call setup_initialGWD()
+		ncid_topoGWD=>topoGWD_file_get_id()
+                call infld('SGH'  ,ncid_topoGWD,dim1name,dim2name, 1,pcols,begchunk,&
+                                endchunk,  var, found, gridname='physgrid')
+                call infld('SGH30',ncid_topoGWD,dim1name,dim2name, 1,pcols,begchunk,&
+                                endchunk,  var30, found, gridname='physgrid')
+                call infld('OC', ncid_topoGWD,dim1name,dim2name, 1,pcols,begchunk,  &
+                                endchunk,  oc,  found, gridname='physgrid')
+                !keep the same interval of OA,OL
+                call infld('OA', ncid_topoGWD,dim1name,'nvar_dirOA',dim2name,1,pcols,1,nvar_dirOA,begchunk, &
+                                endchunk,  oadir(:,:,:),  found, gridname='physgrid')
+                call infld('OL', ncid_topoGWD,dim1name,'nvar_dirOL',dim2name,1,pcols,1,nvar_dirOL,begchunk, &
+                                endchunk,  ol, found, gridname='physgrid')
+                if(.not. found) call endrun('ERROR: GWD topo file readerr')
+                !
+                call close_initial_fileGWD()
+  endif
```
weird indentation here too

whannah1 commented on components/eam/src/physics/cam/comsrf.F90:
```diff
@@ -134,4 +140,32 @@ subroutine initialize_comsrf
     end if
   end subroutine initialize_comsrf
 
+  subroutine initialize_comsrf2
+  use cam_control_mod,  only: ideal_phys, adiabatic
```
Why exactly do we need this routine? Couldn't these variables be initialized in the other initialize_comsrf routine?

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -742,4 +746,1472 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+  !in case the angle is not into the judgement
+  theta=theta1
+  !transform of angle into first quadrant
+  if      (theta1.ge.  0._r8.and.theta1.lt. 90._r8) then
+  theta=theta1
+  else if (theta1.gt. 90._r8.and.theta1.lt.180._r8) then
+  theta=(180._r8-theta1)
+  else if (theta1.gt.180._r8.and.theta1.lt.270._r8) then
+  theta=(theta1-180._r8)
+  else if (theta1.gt.270._r8.and.theta1.lt.360._r8) then
+  theta=(360._r8-theta1)
+  else if (theta1.eq.90._r8.or.theta1.eq.270._r8) then
+  theta=90._r8
+  else if (theta1.eq.0._r8.or.theta1.eq.180._r8) then
+  theta=0._r8
+  endif
+
+  !get dxy
+  if   (theta.ge. 0._r8.and.theta.lt.atan2(dy,dx)/rad) then
+  dxy=dx/cos(theta*rad)
+  else if (theta.ge.atan2(dy,dx)/rad.and.theta.le.90._r8)then
+  dxy=dy/sin(theta*rad)
+  endif
```
another indentation issue, but also this routine could benefit from a description of what it's calculating. It's not immediately clear to me what "dxy" is. I'm guessing it's like a solid angle sort of thing with units of radians squared?

jinboxie replied:
dxy is the model grid length (in meter) in the wind direction. It is added in description.

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -742,4 +746,1472 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+
+!==========================================================================
+
+subroutine grid_size(state, grid_dx, grid_dy)
+  ! Determine the size of the grid for each of the columns in state
+
+  use phys_grid,       only: get_area_p
+  use shr_const_mod,   only: shr_const_pi
+  use physics_types,   only: physics_state
+  use ppgrid,          only: pver, pverp, pcols
+
+  type(physics_state), intent(in) :: state
+  real(r8), intent(out)           :: grid_dx(pcols), grid_dy(pcols)   ! E3SM grid [m]
+
+  real(r8), parameter :: earth_ellipsoid1 = 111132.92_r8 ! World Geodetic System 1984 (WGS84) 
+                                                         ! first coefficient, meters per degree longitude at equator
+  real(r8), parameter :: earth_ellipsoid2 = 559.82_r8 ! second expansion coefficient for WGS84 ellipsoid
+  real(r8), parameter :: earth_ellipsoid3 = 1.175_r8 ! third expansion coefficient for WGS84 ellipsoid
+  real(r8) :: mpdeglat, column_area, degree, lat_in_rad
+  integer  :: i
+  
+  do i=1,state%ncol
+      ! determine the column area in radians
+      column_area = get_area_p(state%lchnk,i)
+      ! convert to degrees
+      degree = sqrt(column_area)*(180._r8/shr_const_pi)
+
+      ! convert latitude to radians
+      lat_in_rad = state%lat(i)*(shr_const_pi/180._r8)
+
+      ! Now find meters per degree latitude
+      ! Below equation finds distance between two points on an ellipsoid, derived from expansion
+      !  taking into account ellipsoid using World Geodetic System (WGS84) reference 
+      mpdeglat = earth_ellipsoid1 - earth_ellipsoid2 * cos(2._r8*lat_in_rad) + earth_ellipsoid3 * cos(4._r8*lat_in_rad)
+      grid_dx(i) = mpdeglat * degree
+      grid_dy(i) = grid_dx(i) ! Assume these are the same
+  enddo
+end subroutine grid_size
```
From back when I implemented the physgrid infrastructure I know we already calculate this elsewhere. Would it be possible to use that area calculation for consistency?

jinboxie replied:
Not sure. I know this is an estimate. But using other ways of calculation may change a little results. Depends on how different your calculation is from this calculation.

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -742,4 +746,1472 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+      !maintain (oa+2) greater than or equal to 0
+      efact    = max(oa1(i)+2._r8,0._r8) ** (ce*fr(i)/frc)
+      efact    = min(max(efact,efmin),efmax)
+!!!!!! cleff (effective grid length) is highly tunable parameter
+!!!!!! the bigger (smaller) value produce weaker (stronger) wave drag
+      cleff    = sqrt(dxy(i)**2._r8 + dxyp(i)**2._r8)
```
do we really need the extra exclamation points here?

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -742,4 +746,1472 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+  real(r8), dimension(its:ite,kts:kte)     :: utendwave,vtendwave,thx,thvx,za
+  real(r8), dimension(its:ite)             :: govrth
+  real(r8), dimension(its:ite,kts:kte+1)   :: zq
+  real(r8)                                 :: tauwavex0,tauwavey0,XNBV,density,tvcon,hpbl2
```
XNBV is a pretty non-descriptive variable name. Looking below it seems to be a temporary work variable - could we use a name to communicate that better?

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -742,4 +746,1472 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+        enddo
+      enddo
+
+      do k = kts,kte
+        do i = its,ite
+          za(i,k) = 0.5_r8*(zq(i,k)+zq(i,k+1))
+        enddo
+      enddo
+    ENDIF
+
+    DO i=its,ite
+      IF (xland1(i) .gt. 0..and.2._r8*var(i).gt.0) then
+        a1=0.00026615161_r8*var(i)**2_r8
+        a2=a1*0.005363_r8
+        DO k=kts,kte
```
Why are you switching between all-capital and all-lowercase letters? Was this all copied and pasted from different sources?

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -742,4 +746,1472 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+
+    DO i=its,ite
+      IF (xland1(i) .gt. 0..and.2._r8*var(i).gt.0) then
+        a1=0.00026615161_r8*var(i)**2_r8
+        a2=a1*0.005363_r8
```
I see you're using a few "magic numbers" in this section. Do these come from a paper or something else we could cite here?

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -742,4 +746,1472 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+        a2=a1*0.005363_r8
+        DO k=kts,kte
+          wsp=SQRT(u1(i,k)**2_r8 + v1(i,k)**2_r8)
+          ! alpha*beta*Cmd*Ccorr*2.109 = 12.*1.*0.005*0.6*2.109 = 0.0759 
+          utendform(i,k)=-0.0759_r8*wsp*u1(i,k)* &
+                         EXP(-(za(i,k)/1500._r8)**1.5_r8)*a2*za(i,k)**(-1.2_r8)*ss_taper
+          vtendform(i,k)=-0.0759_r8*wsp*v1(i,k)* &
+                         EXP(-(za(i,k)/1500._r8)**1.5_r8)*a2*za(i,k)**(-1.2_r8)*ss_taper
```
more "magic numbers" here - any idea where they come from?

whannah1 commented on components/eam/src/physics/cam/gw_common.F90:
```diff
@@ -742,4 +746,1472 @@ subroutine gw_drag_prof(ncol, ngwv, src_level, tend_level, do_taper, dt, &
+     enddo
+   enddo
+  !                                                                       
+  !  limit de-acceleration (momentum deposition ) at top to 1/2 value 
+  !  the idea is some stuff must go out the 'top'                     
+  !                                                                       
+
+   do klcap = lcap,kte
+     do i = its,ite
+       taud_ls(i,klcap) = taud_ls(i,klcap) * factop
+       taud_bl(i,klcap) = taud_bl(i,klcap) * factop
+     enddo
+   enddo
```
It seems lcap is hard coded to be equal to kte, is that intentional?

jinboxie replied:
It is hardcoded, intentional.

whannah1 commented on components/eam/bld/namelist_files/namelist_defaults_eam.xml:
```diff
@@ -126,14 +126,13 @@
 <bnd_topo hgrid="ne16np4"  npg="0">atm/cam/topo/USGS-gtopo30_ne16np4_16xconsistentSGH.c20160612.nc</bnd_topo>
 <bnd_topo hgrid="ne16np4"  npg="2">atm/cam/topo/USGS-gtopo30_ne16np4pg2_16xdel2_20200527.nc</bnd_topo>
 <bnd_topo hgrid="ne30np4"  npg="0">atm/cam/topo/USGS-gtopo30_ne30np4_16xdel2-PFC-consistentSGH.nc</bnd_topo>
-<bnd_topo hgrid="ne30np4"  npg="2">atm/cam/topo/USGS-gtopo30_ne30np4pg2_x6t-SGH.c20210614.nc</bnd_topo>
```
let's not change this default for now and wait until we want this feature to be on by default

jinboxie replied:
Hi Walter, so do I just leave it as default, or do you feel like adding something like a configuration that uses the new topo designed for the oro drags?

whannah1 replied:
I was thinking we should just omit this while it's in stealth mode, but I guess we need something to allow for testing, so let's add this ncdata exception to the testmod that you created.

jinboxie replied:
Hi Walter, is there a command that would just use alternative new topos (depending on various resolutions) that I can add to the testmods?

jinboxie commented:
Hi Walter @whannah1 ,I've made a newer version of the code. The orographic drag toolkit is taken out and will be in a new PR of purely toolkit as you have mentioned. Please check at your convenience.

whannah1 commented:
@wlin7 @rljacob @jinboxie 
I just pushed commits to refactor of how these schemes are implemented to address a few issues
- failing ne4 tests due to wrong topo file being specified - tests only crashed at time step 10 on chrysalis (no PM) so Ln5 tests missed the problem
- the new oro drag data was being added to the physics state, which is not good practice, so I switched these to be stored in pbuf, which also seems to have sped up the tests a small amount (perhaps no significant)
- I moved the initialization into dedicated routines within the "od_common" module to further distinguish the new schemes from the current ones

This helped reduce the number of files being touched. I also included several cosmetic fixes (mostly white space) and made some variable names more verbose. There's several other minor things like this that I might push for fixing or polishing but I don't want to spend any more time on this, and since it's a stealth feature I'm ready to to approve and get it merged.

Does that sound like a good plan? Or would you prefer to have someone else review the PR now that I've added significant changes?

jinboxie commented:
> @wlin7 @rljacob @jinboxie I just pushed commits to refactor of how these schemes are implemented to address a few issues
> 
>     * failing ne4 tests due to wrong topo file being specified - tests only crashed at time step 10 on chrysalis (no PM) so Ln5 tests missed the problem
> 
>     * the new oro drag data was being added to the physics state, which is not good practice, so I switched these to be stored in pbuf, which also seems to have sped up the tests a small amount (perhaps no significant)
> 
>     * I moved the initialization into dedicated routines within the "od_common" module to further distinguish the new schemes from the current ones
> 
> 
> This helped reduce the number of files being touched. I also included several cosmetic fixes (mostly white space) and made some variable names more verbose. There's several other minor things like this that I might push for fixing or polishing but I don't want to spend any more time on this, and since it's a stealth feature I'm ready to to approve and get it merged.
> 
> Does that sound like a good plan? Or would you prefer to have someone else review the PR now that I've added significant changes?

@whannah1 Hi Walter. Thank you for the time contributed to helping this code improve! The plan is great.

rljacob commented:
A few things the PR doesn't make clear.
Its adds 4 new schemes so there's a total of 5 if we include the original one?
This is BFB and STEALTH since the original schemes is left on?
Are the 4 schemes each switchable on/off?  It looks like the 10-year sim was done with all of them on?

whannah1 commented:
> Are the 4 schemes each switchable on/off?

Yes.

rljacob commented:
Notes:  atm group will review results.   Want group lead approval

crterai commented on components/eam/docs/tech-guide/orodrag.md:
```diff
@@ -0,0 +1,35 @@
+
+## Overview
+
+The orographic drag schemes includes two main options: the default Gravity Wave Drag schemeof McFarlane (1987)[@mcfarlane_the_1987] and a new suite of orographic drag parameterization schemes. The new suite includes 4 components all combined in one module (i.e. subroutine gwdo2d). The schemes include orographic gravity wave drag (oGWD, Xie et al.,2020)[@xie_an_2020], flow-blocking drag (FBD, Xie  ...
```
@jinboxie - Would you state in this paragraph, maybe the last sentence, that the default configuration in E3SMv3 has the oGWD of McFarlane (1987) turned on, but all other schemes turned off? 
Also, there is a typo in the first sentence `schemeof` should be `scheme of`.

crterai commented on components/eam/docs/tech-guide/orodrag.md:
```diff
@@ -0,0 +1,35 @@
+
+### Default oGWD scheme
+
+The current default oGWD scheme in E3SMv3 is from McFarlane (1987)[@mcfarlane_the_1987]. It is a linear orographic gravity wave drag scheme that parameterizes the subgrid process of vertical propagation of gravity wave originated from orographic source. The output from this scheme is a vertical profile of drag (or deceleration terms) when gravity wave breaks at higher levels and deposit momemtum  ...
```
Note that it is turned on in E3SMv3.0. 

Grammatical fixes: 
Replace `originated from` with `originating from`
Replace `and deposit momentum` with `and deposits momentum` 
Replace `have impact on` with `impact` (or better yet replace with `improve`, if that is true)

rljacob replied:
You want something more then "The current default oGWD scheme in E3SMv3 is from McFarlane " ?

jinboxie replied:
I think Chris just wants to note that this scheme is turned on by default in E3SMv3.0

crterai replied:
Jinbo's recent revisions have addressed my comment about just noting what is turned on and off by default in E3SMv3.0.

crterai commented on components/eam/docs/tech-guide/orodrag.md:
```diff
@@ -0,0 +1,35 @@
+
+### Default TMS scheme
+
+The turbulent mountain stress (TMS) scheme parameterizes is documented on Richter et al.,(2010)[@richter_the_2010]. It parameterizes the turbulent scale form drag (TOFD) through adding enhanced effective roughness length to the model. This method is justified by the observation that area-averaged wind sufficiently far above hilly terrain behave logarithmically. It leads to significant decrease of ...
```
Suggest removing `parameterizes` in the first sentence.

crterai commented on components/eam/docs/tech-guide/orodrag.md:
```diff
@@ -0,0 +1,35 @@
+
+### New oGWD scheme
+
+The Xie et al.(2020)[@xie_an_2020] oGWD scheme is a nonlinear orographic gravity wave drag scheme that parameterizes the subgrid oGWD (>5km). It includes both the linear oGWD like that of McFarlane (1987)[@mcfarlane_the_1987] and nonlinear oGWD like low-level wave breaking (LLWB). The LLWB downstream often results in downslope windstorm that can enhance the drag by several times that of linear oG ...
```
Note at the end of this description and below that they are all turned OFF in E3SMv3.0.

crterai replied:
The description above suffices so resolving this comment.

crterai commented on components/eam/docs/user-guide/namelist_parameters.md:
```diff
@@ -156,3 +156,20 @@
+| `effgw_oro`        | Efficiency associated with orographic gravity waves.                                                                     | `0.375`                |
+| `tms_orocnst`      | Turbulent mountain stress parameter used when TMS calculation is turned on             | `1.0`                |
+| `tms_z0fac`        | Factor determining z_0 from orographic standard deviation [ no unit ] for TMS.                                                   | `0.75`                |
+| `use_od_ls`        | This namelist controls the new nonlinear oGWD, if used, the nonlinear oGWD is turned on. use_od_ls should not be used at the same time with use_gw_oro.                  | `true`                |
```
Default value should be `false` since it's going in as a stealth feature. We will update this to `true` if and when we turn it on by default. Same goes with `use_od_bl`, `use_od_ss`, and `use_od_fd`

crterai commented on components/eam/docs/tech-guide/orodrag.md:
```diff
@@ -0,0 +1,35 @@
+
+## Overview
+
+The orographic drag schemes includes two main options: the default Gravity Wave Drag scheme of McFarlane (1987)[@mcfarlane_the_1987] and a new suite of orographic drag parameterization schemes. The new suite includes 4 components all combined in one module (i.e. subroutine gwdo2d). The schemes include orographic gravity wave drag (oGWD, Xie et al.,2020)[@xie_an_2020], flow-blocking drag (FBD, Xie ...
```
I didn't catch this the first time around:
`The conceptual location of each scheme is illustrate below.`
should be
`The conceptual location of each scheme is illustrated below.`

And suggest rewording 
`Currently, only the scheme of McFarlane (1987) is opened as default in E3SMv3.0.`
to
`Currently, only the scheme of McFarlane (1987) is turned on by default in E3SMv3.0.`

jinboxie replied:
Hi Chris, I may try to change to "The concept of each scheme is illustrate below." How is that?

crterai commented on components/eam/docs/tech-guide/orodrag.md:
```diff
@@ -0,0 +1,35 @@
+
+### Default oGWD scheme
+
+The current default oGWD scheme in E3SMv3.0 is from McFarlane (1987)[@mcfarlane_the_1987]. It is a linear orographic gravity wave drag scheme that parameterizes the subgrid process of vertical propagation of gravity wave originating from orographic source. The output from this scheme is a vertical profile of drag (or deceleration terms) when gravity wave breaks at higher levels and deposits momem ...
```
Replace 
`This scheme is shown to have improve`
with
`This scheme is shown to improve`

rljacob commented:
@wlin7 you can start merging this.

wlin7 commented:
Merged to next.

whannah1 commented:
e3sm_mmf_integration passes on chrysalis

wlin7 commented:
Re-merged to next.
